### ISSUE TITLE
Files got corrupted

### ISSUE THREAD

**Issue by JimmyBuzzard** [hbons/SparkleShare#1738]
I am using SparkleShare on my own server with encryption option. Today some of the files got corrupted and I am unable to open them in their editors. Restoring files from history does not help. 
Opening it in text editor shows what seems to be base64 encoded characters, 64 per line.  

Has anyone experienced a similar problem?  

EDIT: using version 1.5.0 and the logs do not show anything out of the ordinary  

EDIT: So the file looks like aes-256-cbc encrypted file.  
Optimistically tried decrypting it with the following command:
```openssl aes-256-cbc -d -a -in file.in -out file.out```  
and using the password used for repository encryption, but that gave me a bad decrypt.

**Comment by jschaul** [comment]
what is your version of openssl? (run `openssl version`) In case it is >= 1.1, read on (otherwise your issue is a different one)

Current version of Sparkleshare seems to be incompatible with openssl >= 1.1.0 due to https://stackoverflow.com/questions/39637388/encryption-decryption-doesnt-work-well-between-two-different-openssl-versions/39641378#39641378
The problem occurs when the encrypted repo was created with a version of openssl < 1.1 (it uses md5 to generate a key from your password) and then you try to decrypt it with a version >= 1.1 (where it uses sha256 to generate a key from your password - this key doesn't match the original key so decryption fails).

Allowing Sparkleshare to take the extra backwards-compatible command line option `-md md5` would solve that issue. Ideally this would be configurable via config file and/or guessed depending on the version of openssl used.

Currently the only workaround seems to be to use a version < 1.1 of openssl.

**Comment by JimmyBuzzard** [comment]
> what is your version of openssl? (run openssl version) In case it is >= 1.1, read on (otherwise your issue is a different one)  

OpenSSL 1.1.0c  10 Nov 2016  

What is strange is that I am seeing an encrypted file when I open it in text editor. Are you saying that SparkleShare failed to decrypt the file and left it unchanged?  

The other issues is this doesn't happen always. I replaced the file with a backup and now it seems to work, don't know for how long.

**Comment by hbons** [comment]
Really OpenSSL? :(

This is pretty bad.  @jschaul do you know if there is a way to force enc/dec with the same password algorithm?

**Comment by jschaul** [comment]
@JimmyBuzzard - an issue will occur if you e.g. synchronize files between two clients, where one client runs openssl >= 1.1 and one client runs openssl < 1.1. If both of your computers have the same version of openssl, and you created a new, fresh repository, things will work fine. 

In general this is a compatibility problem however.

@hbons there is the `-md` flag, e.g. `-md md5` (default in openssl < 1.1.0) or e.g. `-md sha256` (default in openssl >= 1.1.0). It's possible to encrypt/decrypt successfully across different openssl versions when specifying this parameter; the change in openssl with version 1.1.0 was the *default* value only. See my example PR https://github.com/hbons/SparkleShare/pull/1740 - but that is more a workaround (and I didn't test this code!) than a solution. I guess it depends how you gauge the importance of backwards compatibility.

**Comment by JimmyBuzzard** [comment]
> an issue will occur if you e.g. synchronize files between two clients, where one client runs openssl >= 1.1 and one client runs openssl < 1.1.

Yes, I can confirm this is the case, the other machine is using `OpenSSL 1.0.2g  1 Mar 2016`, and the problem was replicated on a new file.

**Comment by jschaul** [comment]
@JimmyBuzzard for your specific case, you *might* be able to get your setup working by, on your computer that runs openssl 1.0.2, changing the smudge/clean lines inside `~/Sparkleshare/<your-repo>/.git/config` file to include `-md sha256`, i.e. change 

```
[filter "encryption"]
	smudge = openssl enc -d -aes-256-cbc -base64 -S <some hash value> -pass file:.git/info/encryption_password
	clean = openssl enc -e -aes-256-cbc -base64 -S <some hash value> -pass file:.git/info/encryption_password
```
to 
```
[filter "encryption"]
	smudge = openssl enc -d -aes-256-cbc -base64 -S <some hash value> -pass file:.git/info/encryption_password -md sha256
	clean = openssl enc -e -aes-256-cbc -base64 -S <some hash value> -pass file:.git/info/encryption_password -md sha256
```

after that, new file changes should propagate correctly. Note that this is a very fragile setup - you will end up with a git repo where, on the server, some files will have been encrypted using a md5 based key derivation of your password, and some files will have been encrypted using a sha256 based key derivation. Any third client checking out this repo will be unable to decrypt half your files. 

It is probably a much better idea to try to use the same version of openssl on both machines, and create fresh repos.

**Comment by hbons** [comment]
Fixed for future setups with #1740.
